jax/_src/pallas/core.py

```diff
@@ -722,18 +728,144 @@ def _make_scratch_aval(self, obj: object) -> jax_core.AbstractValue:
   def _make_scalar_ref_aval(self, aval):
     assert False  # Not needed in GridSpec
 
-  def unzip_dynamic_grid_bounds(
-      self,
-  ) -> tuple[GridSpec, tuple[Any, ...]]:
-    static_grid = tuple(
-        d if isinstance(d, int) else None for d in self.grid
-    )
-    dynamic_bounds = tuple(d for d in self.grid if not isinstance(d, int))
-    # We can't use dataclasses.replace, because our fields are incompatible
-    # with __init__'s signature.
-    static_self = copy.copy(self)
-    static_self.grid = static_grid  # type: ignore
-    return static_self, dynamic_bounds
+def unzip_dynamic_grid_bounds(
+    grid_spec: GridSpec) -> tuple[GridSpec, tuple[Any, ...]]:
+  static_grid = tuple(
+      d if isinstance(d, int) else None for d in grid_spec.grid
+  )
+  dynamic_bounds = tuple(d for d in grid_spec.grid if not isinstance(d, int))
+  # We can't use dataclasses.replace, because our fields are incompatible
+  # with __init__'s signature.
+  static_self = copy.copy(grid_spec)
```

Comment:
We could, but we don't need to. Now it is used very little internally, and it is turned into ` GridMapping` very early, which is frozen. `GridSpec` is part of the API, and today making it frozen would break, e.g., aqt2, because their code to create GridSpec relies on refining it. Of course, they could use `replace` ...